### PR TITLE
feat: Add concept of 'revisions' to disruption changes

### PR DESCRIPTION
**Pull request by losvedir**
#### Summary of changes
**Asana Ticket:** [🏹 Arrow uses new schema](https://app.asana.com/0/584764604969369/1185700222235211)

Adds the database support for the "published" vs "draft" feature. For now, every change is automatically published, so current app behavior is the same. However, behind the scenes it will now be storing each change as a DisruptionRevision and a cloned set of associated relationships.

Per discussion, we removed some of the time-based validations, since we've found that we sometimes actually _do_ want to make retroactive changes, but have left a lot of the "obviously wrong" checks, like exception dates falling within the date range and on relevant days of the week.

#### Reviewer Checklist
- [ ] Meets ticket's acceptance criteria
- [ ] Any new or changed functions have typespecs
- [ ] Tests were added for any new functionality (don't just rely on Codecov)
- [ ] This branch was deployed to the staging environment and is currently running with no unexpected increase in warnings, and no errors or crashes.

### CODE REVIEW

File: lib/arrow_web/controllers/api/disruption_controller.ex [hunk 1]
```diff
@@ -1,36 +1,52 @@
...
+
+    data =
+      query
+      |> Repo.all()
+      |> Repo.preload([:adjustments, :days_of_week, :exceptions, :trip_short_names])
+
+    # IO.inspect(data)
+    # require IEx
+
+    # IEx.pry()
```
Comment (lemald) [inline lib/arrow_web/controllers/api/disruption_controller.ex:26]: debugging code

File: lib/arrow/disruption_revision.ex [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+          trip_short_names: [TripShortName.t()] | Ecto.Association.NotLoaded.t(),
+          adjustments: [Arrow.Adjustment.t()] | Ecto.Association.NotLoaded.t(),
+          inserted_at: DateTime.t() | nil,
+          updated_at: DateTime.t() | nil
+        }
+
+  schema "disruption_revisions" do
+    field :end_date, :date
+    field :start_date, :date
+    field :is_active, :boolean
```
Comment (fjlanasa) [inline lib/arrow/disruption_revision.ex:27]: does this have a default, or is it implicitly false?

File: test/arrow/disruption_test.exs [hunk 3]
```diff
@@ -1,774 +1,403 @@
...
-               disruption
-               |> Disruption.changeset_for_delete(@current_time)
-               |> Repo.delete()
+      Repo.update!(Ecto.Changeset.change(d, %{published_revision_id: dr.id}))
 
-      assert changeset.errors == [
-               start_date: {"can't be deleted when start date is in the past", []}
-             ]
+      assert {:ok, dr} = Arrow.Disruption.delete(dr.id)
+      assert Repo.all(from(dr in Arrow.DisruptionRevision, select: count(dr.id))) == [2]
```
Comment (fjlanasa) [inline test/arrow/disruption_test.exs:399]: is the original `DisruptionRevision` marked as `is_active: false` in this case?

File: test/arrow/disruption_test.exs [hunk 4]
```diff
@@ -1,774 +1,403 @@
...
+               %{detail: "Adjustments should have at least 1 item(s)"}
+             ] = ArrowWeb.Utilities.format_errors(e)
     end
+  end
 
-    test "Can't delete exception date that's in the past" do
-      start_date = ~D[2020-01-01]
-      end_date = ~D[2020-01-20]
-      disruption = build_disruption(%Arrow.Disruption{start_date: start_date, end_date: end_date})
+  describe "update/2" do
```
Comment (fjlanasa) [inline test/arrow/disruption_test.exs:163]: what happens to the `is_active` field on the copied `DisruptionRevision`?

**Comment by losvedir** [comment]
Seeing your questions here, I think `is_active` is a bit confusing. It essentially means `is_deleted: false`, the way we're using it. So, previous disruption revisions are correctly left as `is_deleted: false`. It *doesn't* mean that's the revision that is currently serving as the "draft" or "published" revision or something like that. Its genesis, I think, came from my original proposed "statuses" of `pending_deactivation`, `inactive`, `pending_activation`, `active`, but those ended up being unnecessary other than the one use of flagging a revision as deleted.

Should we consider renaming this field? Is `is_deleted` any better? I want to tease apart the meanings of "is this the revision that's being used" vs. "does this revision indicate that the disruption has been / will be deleted".

And regarding the questions of defaults, if we do change it, it's probably worth adding a postgres default of it not being deleted.

File: lib/arrow/disruption_revision.ex (hunk 2 above)
Comment (lemald) [reply lib/arrow/disruption_revision.ex:27]: FWIW the migration does set it to default to true in Postgres.

**Comment by lemald** [comment]
If I recall correctly @digitalcora had some thoughts on the `is_deleted` vs. `is_active` issue, but I don't remember exactly why we went with `is_active`.

File: lib/arrow/disruption/day_of_week.ex [hunk 5]
```diff
@@ -24,7 +24,7 @@ defmodule Arrow.Disruption.DayOfWeek do
   @spec changeset(t(), map()) :: Ecto.Changeset.t()
   def changeset(day_of_week, attrs) do
     day_of_week
-    |> cast(attrs, [:id, :day_name, :start_time, :end_time])
```
[the remark this answers is not in the document]
Comment (losvedir) [reply lib/arrow/disruption/day_of_week.ex:27]: I don't think we do. It works without it. In fact, I believe I had to remove that to *make* it work. That said, from what I semantically understand `cast` to mean, is that these are the parameters we want the user to be able to set. We *don't* want them to set the `id`, so I feel like it doesn't belong in the list. I don't remember why we added it initially, but if it's not necessary now (which from testing and experimentation it isn't), I think it's good to remove.

**Comment by losvedir** [comment]
Alternatively, we could retain `is_active` and have it mean something along the lines of "currently serving as the revision for a (published? draft?) disruption". Then we'd have to change the update code to set previous versions to `false` as necessary. However, I like that less because multiple disruption revisions are "active" for different views at the same time. It also seems to me denormalized in what is canonically the revision for a view - is it the one with the `is_active` flag set a certain way or is it the latest one for a disruption?

**Comment by lemald** [comment]
> Alternatively, we could retain `is_active` and have it mean something along the lines of "currently serving as the revision for a (published? draft?) disruption". Then we'd have to change the update code to set previous versions to `false` as necessary. However, I like that less because multiple disruption revisions are "active" for different views at the same time. It also seems to me denormalized in what is canonically the revision for a view - is it the one with the `is_active` flag set a certain way or is it the latest one for a disruption?

Yeah, I share your hesitation about this approach - I would worry that it would be hard to ensure that that sort of denormalization is kept in sync properly.

**Comment by fjlanasa** [comment]
> > Alternatively, we could retain `is_active` and have it mean something along the lines of "currently serving as the revision for a (published? draft?) disruption". Then we'd have to change the update code to set previous versions to `false` as necessary. However, I like that less because multiple disruption revisions are "active" for different views at the same time. It also seems to me denormalized in what is canonically the revision for a view - is it the one with the `is_active` flag set a certain way or is it the latest one for a disruption?
> 
> Yeah, I share your hesitation about this approach - I would worry that it would be hard to ensure that that sort of denormalization is kept in sync properly.

Ah, ok. I was conflating a `DisruptionRevision` set to `is_active: true` with it's associated `Disruption` having a `published_revision_id` matching the revision.

**Comment by fjlanasa** [comment]
In that case, I don't have strong feelings about `is_active` vs `is_deleted`

**Comment by digitalcora** [comment]
> If I recall correctly @digitalcora had some thoughts on the `is_deleted` vs. `is_active` issue, but I don't remember exactly why we went with `is_active`.

I think `is_deleted` is equally reasonable, and if it would be clearer in this context then we should use it.

@losvedir I think it would be a viable approach, since we are already dynamically defining "the latest revision" without storing an ID anywhere, to also define "the published revision" as "the latest revision that has `is_published` set to true". The act of publishing a revision would change from a Disruption update to creating a new revision identical to the previous one but with the flag set (it would have to be cleared by default when creating new revisions). This would also allow us to keep track of the history of published revisions, which the `published_revision_id` column notably doesn't allow on its own.

All that said, this could be done separately from this change in a backwards-compatible way that likely wouldn't interfere with other work too much.

File: test/arrow/disruption_test.exs [hunk 6]
```diff
@@ -1,774 +1,403 @@
...
-               |> Disruption.changeset_for_delete(@current_time)
-               |> Repo.delete()
+      Repo.update!(Ecto.Changeset.change(d, %{published_revision_id: dr.id}))
 
-      assert changeset.errors == [
-               start_date: {"can't be deleted when start date is in the past", []}
-             ]
+      assert {:ok, dr} = Arrow.Disruption.delete(dr.id)
+      assert Repo.all(from(dr in Arrow.DisruptionRevision, select: count(dr.id))) == [2]
+      assert dr.is_active == false
```
Comment (fjlanasa) [inline test/arrow/disruption_test.exs:400]: maybe we should add an assertion that the `Disruption.published_revision_id == dr.id`

File: test/arrow/disruption_test.exs [hunk 7]
```diff
@@ -1,774 +1,403 @@
...
+      assert length(dr_ids) == 2
+      new_dr_id = Enum.find(dr_ids, &(&1 != dr_id))
+
+      new_dr =
+        Arrow.DisruptionRevision
+        |> Repo.get!(new_dr_id)
+        |> Repo.preload([:exceptions, :days_of_week, :trip_short_names])
+
+      assert new_dr.disruption_id == d.id
+      assert new_dr.disruption_id == d.id
```
Comment (fjlanasa) [inline test/arrow/disruption_test.exs:201]: duplicate assertion here. maybe we should add an assertion that the Disruption.published_revision_id == dr.id

Comment (losvedir) [reply test/arrow/disruption_test.exs:201]: Oh, good catch on the duplicate. We'll remove.

I thought about the `published_revision_id == id` thing, but while that's currently the case for `Disruption.update`, it won't be going forward. I think we can skip the assertion for now, but I also wouldn't be opposed to adding it, and then updating the test when the behavior changes.

Comment (fjlanasa) [reply test/arrow/disruption_test.exs:201]: ah, ok. fair enough about the extra assertion.

Comment (lemald) [reply test/arrow/disruption_test.exs:201]: I like the idea of testing the current behavior now and just updating the test when we update the behavior.

**Pull request merged**